processor/processor.go

```diff
@@ -2160,35 +2172,73 @@ func (proc *HandleT) pipelineWithPause(ctx context.Context, fn func(ctx context.
 	}
 }
 
+//Since mainpipeline func can process jobs in a single batch (when number of jobs are less)
+// and in multiple batches (when number of jobs are more). So, to identify if the received
+//job is a sub-part of a job is a single job. So, `isSplit` variable is needed to be passed
+//accross go routines.
+type subJobT struct {
+	subJobs  []*jobsdb.JobT
+	haveMore bool
+}
+
+func splitJob(jobs []*jobsdb.JobT) []subJobT {
```

Comment:
@saurav-malani splitJobs function may not respect the `subJobSize` for the last page if the `size` of the jobs is not a multiple of the `subJobSize`.

E.g. if `jobSize` is 3999 and `subJobSize` is 2000, `splitJob` will produce an array with a single element of size 3999, instead of two elements, one with 2000 and another with 1999.

Please consider consulting my example here and also make sure to add the relevant tests that I have included:
https://github.com/rudderlabs/rudder-server/pull/1717#discussion_r823629433